### ISSUE TITLE
Accessibility (text-to-speech) for visually-impaired gamers

### ISSUE THREAD

**Issue by umeboshilover-gamedev** [YoYoGames/GameMaker-Bugs#12236]
### Is your feature request related to a problem?

From a Reddit thread: "In terms of text, every bit of text in your game is a sprite, not a string of text displayed through a font. I dunno how most (or any really) screen readers work, but unless they are performing visual analysis of the screen I'm 99% sure they aren't going to be able to extract text information to read from a GM game."

### Describe the solution you would like to see

Dear GameMaker Studio and Github.

I was made aware of this lacking feature by an accessibility consultant. 

I am an emerging, Deaf game developer. I'm trying to make a Deaf-accessible games, but recently I've considered making games blind-accessible as well.
However, to my knowledge, GameMaker can't facilitate this feature.

Godot 4.5 has recently implemented full screen reader compatibility. Developers can code their games so that blind and visually-impaired gamers' screen readers read on-screen text, like buttons or dialogue texts. Users only need to have their screen reader turned on and Godot will detect that, so it doesn't need to be activated manually.

Now, blind and visually-impaired players can play Godot games, as long as the developers account for their needs. Some examples: Necromancer Nonsense, The Castle and Lacus Opportunitas [with links below] which got 10th, 3rd and 1st place in the Games for Blind Gamers Jam 4, in 2025 on itch.io.
https://ericbomb.itch.io/necromancer-nonsense
https://punishedfelix.itch.io/the-castle
https://shiftbacktick.itch.io/lacus-opportunitas

We would like it if GameMaker could follow Godot's steps. You can take a look at their Github to understand which needs need to be met for assistive technology compatibility. 

Blind and visually-impaired gamers do play games, and they would love GameMaker to welcome them and set the difference in a vastly innacessible world.
Godot accessibility Github: https://github.com/lightsoutgames/godot-accessibility

Kind regards.

**Comment by cioran0** [comment]
I'm working with TTS for my android app and mildly familiar w/STT tools since I'm also a TTv5 Section 508 Trusted Tester and occasionally do some a11y work. 

I looked up Godot's implementation. I would use a similar one if it were me. They're FOSS.

It's as follows -

speechD (HLI, same for all) --> Voice/etc
https://freebsoft.org/speechd

Use this for the HLI.

Here are the voice/etc per OS

Windows - Microsoft Speech SDK 5.1 
https://www.microsoft.com/en-us/download/details.aspx?id=10121

MS is pushing corporate clients to use AI tools but quietly updates an older (and free) solution for use with the OS.

Linux - Linux Speech Dispatcher
Default. I would consider using Piper and automate install w/Pied personally. S-D sounds like rubbish IME.
https://github.com/rhasspy/piper

Mac - AVSpeechSynthesizer
https://developer.apple.com/documentation/avfaudio/avspeechsynthesizer

You can also use DNN's instead of these and any number of online solutions, but since they may incur costs, I would want to consider these first.

Cheers,
Mike

**Comment by zersiax** [comment]
Wouldn't it be easier, at least on Windows, to talk directly to screen readers using an existing library? Can GameMaker FFI to , say, Tolk for example? https://github.com/dkager/tolk
Similar libraries exist in Python, c#, others...

**Comment by cioran0** [comment]
@zersiax  Interesting IDK that one! Might be a neat choice for windows. They should check it out! 

I went with a design framework modeled on pre-existing implementation in a game engine (Godot) that works already. I haven't used this feature of Godot, but if you can weigh in I'd be interested. @umeboshilover-gamedev talks favorably about it up top. And I've heard similar elsewhere.

I'd still probably call SAPI directly in that fashion since it's less likely to break things and only subject to breaking on SAPI updates, not those of multiple screen readers. Which all ultimately use SAPI anyway so they'd all be subject to breaking on SAPI updates (which are rare). Cut out the middleman, so to say.  Also since the Gamemaker people are proprietary, LGPLv3 may be an issue. IDK how much they'd modify the library. Godot is MIT.

But yeah if this works better for them, why not. More options is always good. They are still going to have to write up their own stuff for mac and IDK if they support linux, but at least mac.

I get where calling the screen reader API makes sense for lots of stuff though and it is easy to call. Seems like there's a couple of these, overlapping people and some forks. Universal Speech looks interesting. [Universal Speech link](https://github.com/qtnc/UniversalSpeech
)

**Comment by medeaoblongata** [comment]
Yes. **Please** do not reinvent the wheel. It is far better to lean on the host system accessibility APIs than try to implement a home grown accessibility solution with speech synthesisers and whatnot (which will inevitably fall short for some users).

For example: If you hope to support blind or partially-sighted users by adding speech synthesis, you will be excluding users that rely on braille devices instead of screen readers that use synthesised speech. 

If you use the host system accessibility APIs then you offload all that complexity onto the system vendors, who already support both, and game developers only need to worry about making sure that game objects, roles and states have meaningful text labels. 

Unity has done exactly this. FWIW.
